specs/proposal/571115-provideWidgetToEditAValueInEditionTable/571115-provideWidgetToEditAValueInEditionTable.md

```diff
@@ -0,0 +1,97 @@
+
+## Introduction
+
+The goal of this issue is to be able to provide a specific widget to edit value in an edition table.
+
+Currently, the editor offered to edit a cell depends on the contents of the cell. It can be:
+* a combo-box if the content can only be a value in a pre-defined list (enumeration);
+* a check-box if the content is a boolean;
+* a dialog box for many-valued references
+* a text editor in all the other cases;
+
+![Combo box sample](EditionTableComboBoxWidgetSample.png)
+
+For example, in this table from Family sample, a "Combo box" is available to edit the cells of "Mother" column.
+
+When the content of the column does not correspond to a "simple" feature, this "automatic" behavior is not possible and a simple text editor is proposed.
+
+With this feature, the specifier could provides its own org.eclipse.jface.viewers.CellEditor for a specific feature column.
+
+## Detailed Specification
+
+For this enhancement, Obeo proposes that the specifier could precise a `CellEditorFactory` for each column. This factory will be responsible to return a specific `org.eclipse.jface.viewers.CellEditor` adapted to handle the content of the column. For the end-user, the behavior will be the same than with existing `CellEditor` provided by Sirius for "simple" feature, as explained above.
+
+To provide its CellEditorFactory, the specifier will use a new tool named "Cell Editor". This new tool will be near the “Label Edit” tool.
+![Cell Editor tool menu](CellEditorContextualMenu.png)
+
+A column can have only one "Label Edit" tool or one "Cell Editor" tool. These two tools are exclusive.
+
+A "Cell Editor" tool will have the same variables has a "Label Edit" tool. The current variables, initialized in `org.eclipse.sirius.table.business.internal.metamodel.TableToolVariables.caseLabelEditTool(LabelEditTool)`, for a "Label Edit" tool are:
+* element: The semantic currently edited element (`target` attribute of the current cell).
+* table: The current `DTable`.
+* line: The `DLine` of the current `DCell`.
+* lineSemantic: The semantic element corresponding to the line  (`target` attribute of the current line).
+* root: The semantic root element of the table.
+
+![Cell Editor tool menu](VariablesInLabelEditTool.png)
+
+These variables will be accessible directly in the `CellEditorFactory`. The specifier has the responsibility to return a `CellEditor` compliant with the expected value of the cell.
+The "Cell Editor" tool will have only one property. The qualified class name of the `CellEditorFactory`. This class is a class of the Viewpoint Specification Project, like for [Java Extension](https://www.eclipse.org/sirius/doc/specifier/general/Writing_Queries.html#service_methods) declared the "Viewpoint".
+![Service VSM Registration](serviceVsmRegistration.png)
+
+The `CellEditorFactory` should implements the interface `org.eclipse.sirius.table.tools.internal.command.ITableCellEditorFactory`:
+
```

Comment:
I did not catch the runtime behavior
For me, ITableCellEditorFactory should have a method `getCellEditor(Map<String, Object> parameters, TransactionalEditingDomain ted)`
It is up to the implementor to leverage the parameters and use the ted to execute an EMF Command

Reply:
Digging a little more in `ICellEditor`, I effectively saw `getEditorControl` and `getEditorValue`, but I am not sure to understand, in case of complex widget and hence a complex Control,  how the consumer of this ICellEditor (that is Sirius) will know when to call `getEditorValue`.

Reply:
For the first point, you're partially right. The method should return a CellEditor. But the TED is not necessary. No modification is done by the CellEditor.
For the second point, Sirius uses the CellEditor.getValue() method in DFeatureColumnEditingSupport like in this stack:
	DFeatureColumnEditingSupport$1(DialogCellEditor).doGetValue() line: 274	
	DFeatureColumnEditingSupport$1(CellEditor).getValue() line: 478	
	DFeatureColumnEditingSupport(EditingSupport).saveCellEditorValue(CellEditor, ViewerCell) line: 115	
	TreeViewerEditor(ColumnViewerEditor).saveEditorValue(CellEditor) line: 446	
	TreeViewerEditor(ColumnViewerEditor).applyEditorValue() line: 312	
	ColumnViewerEditor$1.applyEditorValue() line: 165	
	CellEditor$1.run() line: 335	
	SafeRunner.run(ISafeRunnable) line: 45	
	786669171.run(ISafeRunnable) line: not available	
	SafeRunnable.run(ISafeRunnable) line: 174	
	DFeatureColumnEditingSupport$1(CellEditor).fireApplyEditorValue() line: 332	
	DFeatureColumnEditingSupport$1(DialogCellEditor).lambda$0(SelectionEvent) line: 254	
	526379473.accept(Object) line: not available	
	SelectionListener$1.widgetSelected(SelectionEvent) line: 84